### PR TITLE
feature: add panic_with_felt252(felt252) to corelib

### PR DESCRIPTION
**Pull request by Amxx**
Some contract use assert to check a condition, and panic if the condition is not meet. There are cases where the panic is not conditional, or at least not the result of a boolean expression.

```rust
match someFunction(args) {
   A() => doSomething(),
   B() => doSomethingElse(),
   C() => assert(false, 'should not happen')
}
```

The issue with assert(false, code) is that its return type is void, causing possible type conflict with the other branches of the match. Here we want a function that returns never. panic(Array<felt252>) does that, but is not as user friendly as assert because you need to build Array<felt252>.

This PR adds a throw function (open to other names) that uses code_error just like assert but returns never just like panic.

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568)

### CODE REVIEW

**Review by spapinistarkware — CHANGES_REQUESTED** [review]
> Reviewed 7 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568)* status: 7 of 8 files reviewed, 2 unresolved discussions (waiting on @Amxx)

___
*[`corelib/src/lib.cairo` line 379 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2568#-NR3SXtH0wW8R_9O_WAj:-NR3SXtH0wW8R_9O_WAk:bsp2rjv) ([raw file](https://github.com/starkware-libs/cairo/blob/c19df597dd6258a816b27edde974782b626ec650/corelib/src/lib.cairo#L379)):*
> ```Cairo
> 
> #[inline(always)]
> fn throw(err_code: felt252) -> never {
> ```

I'm not sure about the name. In the future we will probably have something better and generic for panic. maybe something like `panic_with_felt`?



_Code quote:_
```Cairo
throw
```
___
*[`corelib/src/serde.cairo` line 109 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2568#-NR3S7Jh2FF00yj2bsP5:-NR3S7Jh2FF00yj2bsP6:b-p59q28) ([raw file](https://github.com/starkware-libs/cairo/blob/c19df597dd6258a816b27edde974782b626ec650/corelib/src/serde.cairo#L109)):*
> ```Cairo
> ) {
>     // TODO(orizi): Replace with simple call once inlining is supported.
>     // This could be: gas::withdraw_gas().expect('Out of gas');
> ```

This shouldn't create different code. Can you do this change, and run the tests with `CAIRO_FIX_TESTS=1`, and push, so I can see the differences?

**Review by spapinistarkware — COMMENTED** [review]
> Reviewed 1 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568)* status: all files reviewed, 2 unresolved discussions (waiting on @Amxx)

**Comment by Amxx** [comment]
> I'm not sure about the name. In the future we will probably have something better and generic for panic. maybe something like `panic_with_felt`?

I'm open to any name.

**Comment by Amxx** [comment]
> Can you do this change, and run the tests with `CAIRO_FIX_TESTS=1`, and push, so I can see the differences?

see [3a23dd3](https://github.com/starkware-libs/cairo/pull/2568/commits/3a23dd3305be57ad7cc3fa05dcd7d8188d0df401)

**Review by spapinistarkware — COMMENTED** [review]
> Reviewed 1 of 3 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568)* status: 8 of 10 files reviewed, all discussions resolved

___
*[`corelib/src/serde.cairo` line 109 at r1](https://reviewable.io/reviews/starkware-libs/cairo/2568#-NR3S7Jh2FF00yj2bsP5:-NR3_FDoBc0USRGi9GEm:bowgsz4) ([raw file](https://github.com/starkware-libs/cairo/blob/c19df597dd6258a816b27edde974782b626ec650/corelib/src/serde.cairo#L109)):*
<details><summary><i>Previously, spapinistarkware (Shahar Papini) wrote…</i></summary><blockquote>

This shouldn't create different code. Can you do this change, and run the tests with `CAIRO_FIX_TESTS=1`, and push, so I can see the differences?
</blockquote></details>

Ye, this is an optimization we are currently working on. You can bring back the TODO, but change it to:
```
TODO(orizi): Replace with ... once "match on enum init" optimization is implemented.
```

**Comment by Amxx** [comment]
> Your line is too long. Keep it under 100 characters please.

Done. Feel free to propose any change directly in github.

**Comment by spapinistarkware** [comment]
We are using Reviewable, so the comments are inlined there. It's the link at the beginning

**Review by orizi — COMMENTED** [review]
> Reviewed 4 of 5 files at r3, 2 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568)* status: 9 of 10 files reviewed, 2 unresolved discussions (waiting on @Amxx and @spapinistarkware)

File: corelib/src/array.cairo [hunk 1]
```diff
@@ -121,14 +121,7 @@ impl ArrayTCloneImpl<T, impl TClone: Clone::<T>, impl TDrop: Drop::<T>> of Clone
...
 ) {
-    match withdraw_gas() {
-        Option::Some(_) => {},
-        Option::None(_) => {
-            let mut data = array_new();
-            array_append(ref data, 'Out of gas');
-            panic(data);
-        },
-    }
+    withdraw_gas().expect('Out of gas');
```
Comment (Amxx) [inline corelib/src/array.cairo:124]: It appears this change causes `Failed calculating gas usage, it is likely a call for `gas::withdraw_gas` is missing.`

```rust
match withdraw_gas() {
    Option::Some(_) => {},
    Option::None(_) => panic_with_felt252('Out of gas'),
}
```
works though

**Comment by Amxx** [comment]
I tried adding 
```rust
impl NeverDrop of Drop::<never>;
```

But that cause issues with `TupleSize1Drop`, `TupleSize2Drop`, `TupleSize3Drop` and `TupleSize4Drop`.

**Review by spapinistarkware — COMMENTED** [review]
Instead of dropping it, match on it.

> Reviewed 1 of 3 files at r6, 1 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568#-:-NRSDdc2EOFHc7Y3oD7u:b-ox9maz)* status: 8 of 10 files reviewed, all discussions resolved (waiting on @orizi)

**Review by spapinistarkware — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r4, 1 of 3 files at r6, 1 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568#-:-NRSECie7Z_49vHoaa0l:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Amxx)

**Review by spapinistarkware — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568#-:-NRcgjAN3T7aMeAbFeO6:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Amxx)

**Review by orizi — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r9, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568#-:-NRiOuCU7rXAlPKZBJ1g:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Amxx)

**Review by orizi — APPROVED** [review]
> Reviewed 4 of 4 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Amxx)

**Review by Amxx — COMMENTED** [review]
I keep seeing "LGTM" message with the message:

> Reviewable status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Amxx)

What is it that I have to do, beside fixing all the conflicts that keep happening ?

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568#-:-NRqfTMADIEwpv2LD0Hs:bnx2xfe)* status: 11 of 12 files reviewed, all discussions resolved (waiting on @orizi)

**Review by spapinistarkware — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r9, 4 of 4 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/2568#-:-NRqgYMz4Ww-hUSSR5Cs:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Amxx)

**Pull request merged**